vllm/v1/sample/logits_processor/__init__.py

```diff
@@ -177,9 +181,125 @@ def build_logitsprocs(
+        self.is_pin_memory = is_pin_memory
+
+        # V0 logits processors can have 2 or 3 arguments
+        assert (nargs := self._nargs) in {2, 3}
+        self.nargs = nargs
+
+    @property
+    @abstractmethod
+    def _nargs(self) -> int:
+        """Returns the number of arguments supported by this V0 logits
+        processor. Must be 2 or 3"""
+        raise NotImplementedError()
+
+    @abstractmethod
+    def _new_logits_processor(
+        self,
+        params: SamplingParams,
+    ) -> Optional[LogitsProcessorV0]:
+        """Consume request info; return a V0-style logits processor.
+
+        Implementer can assume that the vLLM engine config, the hardware
+        accelerator device, and an indication of whether pin memory is
+        available can respectively be found in self.vllm_config, self.device,
+        and self.is_pin_memory
+
+        Return None if logits processor does not need to be applied to request
+
+        Args:
+          params: request sampling params
+
+        Returns:
+          None if logits processor should not be applied to request; otherwise
+          returns a `LogitsProcessorV0` instance
+        
+        """
+        raise NotImplementedError()
+
+    def _new_state(
+        self,
+        params: SamplingParams,
+        prompt_tok_ids: list[int],
+        out_tok_ids: list[int],
+    ) -> Optional[V0WrapperState]:
+        """Return state representation for new request
+
+        Returns None if logits processor is not applicable to request
+
+        Args:
+          params: request sampling params
+          prompt_tok_ids: request prompt token ids
+          out_tok_ids: decoded tokens so far for this request
+
+        Returns:
+          (new logits processor, prompt token ids,
+           previous decoded tokens) or None
+        
+        """
+        if (v0_lp := self._new_logits_processor(params)):
+            return (
+                v0_lp,
+                prompt_tok_ids,
+                out_tok_ids,
+            )
+        return None
+
+    def update_state(self, batch_update: Optional[BatchUpdate]):
+        process_dict_updates(
+            self.req_info,
+            batch_update,
+            self._new_state,
+        )
+
+    def apply(self, logits: torch.Tensor) -> torch.Tensor:
+        if not self.req_info:
+            return logits
+
+        # Apply V0 logits processors to corresponding rows of logits tensor
+        if self.nargs == 2:
+            for req_idx, (v0_lp, _, out_tok_ids) in self.req_info.items():
+                logits[req_idx] = v0_lp(out_tok_ids,
+                                        logits[req_idx])  # type: ignore
+        else:
+            # nargs == 3
+            for req_idx, (v0_lp, prompt_tok_ids,
+                          out_tok_ids) in self.req_info.items():
+                logits[req_idx] = v0_lp(prompt_tok_ids, out_tok_ids,
+                                        logits[req_idx])  # type: ignore
+
```

Comment:
Could we use the apply method suggested here: https://github.com/njhill/vllm/commit/2b205b31fe02c0a4a2b190e66042398cc3e67111#diff-6278b323b85096bd74b396634d6f6edc78a369b4ce50d52dc8f3203ab0d3ada8R54-R61

In particular this does not update the logits tensor if they are already updated in-place.